### DISCUSSION TITLE
Question about cancel safety of request

### DISCUSSION THREAD

**Discussion by hs-vc** [hyperium/hyper#3143] [Q&A] [answered]
Is it safe to reuse the hyper::Client object created for http2 after a request made by it is cancelled?  For example, in the code below, is there any issue with reusing ```client``` after the ```fetch``` is cancelled by timeout in the ```test()```?  I am worried that the HTTP/2 client usually maintains a single connector and that reusing the client object after cancelling a request might result in unexpected behavior or that resources associated with the cancelled request might not be cleaned up properly.

```rust
use hyper::{body::HttpBody as _, Client, http, Uri};
use hyper::client::HttpConnector;
use hyper_tls::HttpsConnector;

type Result<T> = std::result::Result<T, Box<dyn std::error::Error + Send + Sync>>;

async fn fetch(client: &Client<HttpsConnector<HttpConnector>>, url: Uri) -> Result<()> {
    let mut res = client.get(url).await?;

    println!("Response: {}", res.status());
    println!("Headers: {:#?}\n", res.headers());

    // Stream the body, writing each chunk to stdout as we get it
    // (instead of buffering and printing at the end).
    while let Some(next) = res.data().await {
        let chunk = next?;
        println!("{:?}", chunk);
    }
    Ok(())
}


async fn test() {
    let uri: Uri = "https://www.example.com".parse().unwrap();
    let client = Client::builder()
        .http2_only(true)
        .build::<_, hyper::Body>(hyper_tls::HttpsConnector::new());
    
    let _ = tokio::time::timeout(std::time::Duration::from_secs(10), fetch(&client, uri)).await;

    // Reuse the connection after the request may have been timed out
    do_something(client).await
}


async fn do_something(client: Client<HttpsConnector<HttpConnector>>) {
    // Do something with client
}
```

**Comment by seanmonstar** [comment, marked as answer]
Yes, you may reuse the client even after canceling a request. If it were HTTP/1, the connection would be deemed unusable and so new ones would just be made. For HTTP/2, a `RST_STREAM` will be sent to the server that the stream is no longer desired.

**Reply by hs-vc** [reply]
I see, thanks for the explanation.
